test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,338 @@
+  validateAllZero(data_message);
+}
+
+//
+TEST_F(HystrixSinkTest, Disconnect) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  std::string data_message = TestUtility::bufferToString(buffer);
+  EXPECT_EQ(getStreamField(data_message, "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+
+  // Disconnect.
+  buffer.drain(buffer.length());
+  sink_->unregisterConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Reconnect.
+  buffer.drain(buffer.length());
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  data_message = TestUtility::bufferToString(buffer);
+  EXPECT_EQ(getStreamField(data_message, "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+}
+
+TEST_F(HystrixSinkTest, AddAndRemoveClusters) {
+  InSequence s;
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // New cluster.
+  const std::string cluster2_name{"test_cluster2"};
+  ClusterTestInfo cluster2(cluster2_name);
+
+  // Add cluster.
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  cluster_map_.emplace(cluster2_name, cluster2.cluster_);
+  // Redefining since cluster_map_ is returned by value.
+  ON_CALL(cluster_manager_, clusters()).WillByDefault(Return(cluster_map_));
+
+  // Generate data to both clusters.
+  sink_->flush(source_);
+  for (int i = 0; i < 12; i++) {
+    buffer.drain(buffer.length());
+    // Cluster 1
+    ON_CALL(cluster1_.error_5xx_counter_, value()).WillByDefault(Return((i + 1) * 17));
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * 7));
+    cluster1_.cluster_info_->stats().upstream_rq_timeout_.add(1);
+    cluster1_.cluster_info_->stats().upstream_rq_per_try_timeout_.add(2);
+    cluster1_.cluster_info_->stats().upstream_rq_pending_overflow_.add(8);
+
+    // Cluster 2
+    ON_CALL(cluster2.error_5xx_counter_, value()).WillByDefault(Return((i + 1) * 1));
+    ON_CALL(cluster2.retry_5xx_counter_, value()).WillByDefault(Return((i + 1) * 2));
+    ON_CALL(cluster2.error_4xx_counter_, value()).WillByDefault(Return((i + 1) * 3));
+    ON_CALL(cluster2.retry_4xx_counter_, value()).WillByDefault(Return((i + 1) * 4));
+    ON_CALL(cluster2.success_counter_, value()).WillByDefault(Return((i + 1) * 3));
+    cluster2.cluster_info_->stats().upstream_rq_timeout_.add(3);
+    cluster2.cluster_info_->stats().upstream_rq_per_try_timeout_.add(1);
+    cluster2.cluster_info_->stats().upstream_rq_pending_overflow_.add(5);
+
+    sink_->flush(source_);
+  }
+
+  std::unordered_map<std::string, std::string> cluster_message_map =
+      buildClusterMap(TestUtility::bufferToString(buffer));
+  ASSERT_NE(cluster_message_map.find(cluster1_name_), cluster_message_map.end());
```

Comment:
per current ASAN failure, recommend adding 
  << "cluster1_name=" << cliuster1_name_;
to this ASSERT_NE, and similar below.

Reply:
this is still pending.

Reply:
I don't understand

Reply:
My suggestion was not just to fix your ASAN problem, but also generally to make EXPECT/ASSERT failures actionable, should they occur again in the future.

When checking map/set membership in an ASSERT/EXPECT the test infra will show the code and the actuals, but not automatically show the args, so it would be aid debugging to stream in the value as I showed. This is particularly important if the failures may be flaky as they might not show on repro, so you want to structure your EXPECT/ASSERT calls to get something actionable in the log.